Feature/snapshot

timelyportfolio opened the pull request:
# Better Snapshot

Reference: [plotly #83](https://github.com/plotly/plotly.js/issues/83)

## Objectives

1.  Attach a `toImage` method directly to `Plotly`.
2.  Use promises
2.  Allow resize options for the snapshot in the new `toImage` and/or the snapshot modebar button.

### 1) Attach `toImage`

The easiest way to attach `toImage` would be to piggyback off the existing snapshot modebar button, since the snapshot button `click` method best handles this from start to finish.

See [codepen](http://codepen.io/timelyportfolio/pen/ZWvyYM) for how we can currently accomplish this.

```
Plotly.toImage = function(){
  gd._fullLayout._modeBar.buttons.filter(
    function(btn){return btn[0].name==="toImage"
  })[0][0].click(gd)
}
```

~~While this works, this is an unworkable approach.  First, we assume that the modebar exists and contains the snapshot button.  To get around this, perhaps we could generalize the snapshot `click` handler and make it available outside of the modebar context in [snapshot](https://github.com/plotly/plotly.js/tree/master/src/snapshot).  Currently, the click handler is specified in these [lines](https://github.com/plotly/plotly.js/blob/master/src/components/modebar/buttons.js#L50-L94), so its broader usefulness is limited.  In cff741ac186988454fbb99a2c3715f84d7c9b720, I took a first pass at a new `toSnapshot` method in `Plotly.Snapshot`.~~

I misunderstood the objective.  We want to attach `toImage` to `Plotly`, so let's try again.  Wait to do this until the other steps are complete.

### 2) Use promises
I forgot that we want `toImage` to return a promise.  It seems that we should replace the `EV`/`EventEmitter` with `Promises`.  I see two places where this replacement needs to happen in the pipeline [`svgtoimg`](https://github.com/plotly/plotly.js/blob/master/src/snapshot/svgtoimg.js#L12) and [`toimage`](https://github.com/plotly/plotly.js/blob/master/src/snapshot/toimage.js).  I was able to get this "working" in 937340c51d8c9b99df7e5d73f26f9a9bbffbfcd3 and ac869317edda6b6c84ba01fdda3704065782add9.  Although it works, I am not sure the changes follow Plotly's conventions.  Here is a quick example of what we can do now with promises using [codepen example](http://codepen.io/plotly/pen/EPqXWz).

```
var trace1 = {
  y: [5, 4, 3, 2, 1, 0, 1, 2, 3, 4, 5],
  mode: 'markers',
  marker: {
    size: 40,
	 color: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    colorscale: 'Viridis'
  }
};

var data = [trace1];

var layout = {
  title: 'Scatter Plot with a Color Dimension'
};

var img = document.createElement('img');
document.body.append(img);

Plotly.newPlot('myDiv', data, layout)
  .then(Plotly.Snapshot.toImage)
  .then(function(url){img.src=url});
```

### 3) Allow resize

We'll need to do two things to accomplish this task.  One, I think is fairly easy, and requires the ability to pass `height` and `width` in the `opts` argument of `toImage` to `clonePlot`.  760a93bc76da940cefdef94cee5e4eab566efbb7 is my first try at this.  It seems to work.  Second, we'll need to come up with the UI to specify `height` and `width`.  I looked around the Plotly codebase for a modal or similar UI that I could mimic, but I did not find anything.  If we do a modal, it seems like we should put it in `Lib` similar to `notifier`.  At a minimum, the UI should have textareas for `height` and `width`.  If possible, though, it would be nice to get a live updating preview to reflect the changes to the `height` and `width`.  While this is nice, I don't think it is necessary.

timelyportfolio commented:
I just ran the jasmine tests, and the [old tests](https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/snapshot_clone_test.js) pass even with our changes, but this is because the tests only test `clone`.  I'll look for opportunities to test once I collect opinions and thoughts.

etpinard commented on src/snapshot/tosnapshot.js:
```diff
@@ -0,0 +1,74 @@
+    var Lib = Plotly.Lib;
+  
+    // check for undefined opts
+    opts = (opts) ? opts : {};
```
we would usually write `opts = opts || {};` in situations like this one.

timelyportfolio replied:
guessing same for the next line as well

`opts.format = opts.format || 'png';`

etpinard replied:
yes

etpinard commented on src/snapshot/svgtoimg.js:
```diff
@@ -9,62 +9,61 @@
-
-    img.src = url;
-
-    return ev;
```
Unfortunately, making `svgtoimg.js` return a promise instead of an event object would be a breaking change as some adventurous users might have used `Plotly.Snapshot.svgToImg` in their app.

etpinard replied:
@timelyportfolio I should have mentioned :arrow_double_up: to you before, my mistake.

timelyportfolio replied:
no problem, I should have asked.

etpinard commented on src/snapshot/svgtoimg.js:
```diff
@@ -9,62 +9,61 @@
 
-var EventEmitter = require('events').EventEmitter;
-
 function svgToImg(opts) {
```
I think the solution would be to make `svgToImg` return an event object by default and optionally an promise (e.g. in the case of `Plotly.Snapshot`).

timelyportfolio replied:
ok, I'll get that working.  Maybe in `2.0.0` we can move to all promises.  Are you thinking in `opts` something like `{promise:true}`?

etpinard replied:
> Maybe in 2.0.0 we can move to all promises.

That's the goal. We un-exposed `Plotly.Snapshot` in `v2.0.0`.

> Are you thinking in opts something like {promise:true}

Exactly. So that `Plotly.toImage` will call `svgToImg` with `{ promise: true }` under the covers.

etpinard commented on src/snapshot/index.js:
```diff
@@ -34,7 +34,8 @@ var Snapshot = {
     svgToImg: require('./svgtoimg'),
-    toImage: require('./toimage')
+    toImage: require('./toimage'),
+    toSnapshot: require('./tosnapshot')
```
Let's put `toSnapshot` in a new file called `src/plot_api/to_image.js`.

timelyportfolio replied:
actually, `tosnapshot` remains from my first misguided attempt.  I can eliminate, since I'm not sure it has much value any more with our improved promise-based `toimage`.

etpinard replied:
Oops. Scratch  :arrow_double_up: , `toSnapshot` is essentially what the modebar uses. Putting that in a separate file in `src/snapshot/` makes sense. 

Maybe you could rename `tosnapshot.js` -> `download.js` instead?

etpinard commented on src/snapshot/tosnapshot.js:
```diff
@@ -0,0 +1,74 @@
+ * @param {object} opts option object
+ * @param opts.format 'jpeg' | 'png' | 'webp' | 'svg'
+ */
+function toSnapshot(gd, opts) {
```
we should probably expose a _download image_ functionality too through the `Plotly` object.

Maybe:

``` js
Plotly.plot('graph', data, layout)
 .then(Plotly.downloadImage({ /* e.g.: format: 'png', width: 700, height: 500 */ });
```

@timelyportfolio thoughts?

timelyportfolio replied:
I like that idea, and I think it greatly simplifies programatically creating a snapshot.  Also, I think it will be nice to have a single source for this purpose and the modebar.  I'll work on tidying it up, and then we can figure out best name and where to attach.

etpinard commented:
@timelyportfolio I feel like our discussion here and on #83 hasn't put forward any hard requirements for this issue, let me write them down here:
- Add a `Plotly.toImage` (official) method. This method is essentially the current `Plotly.Snapshot.toImage` but with options handlers for `format` (either 'svg', 'png', 'jpg' and 'webp'), `width` (which should default to the current `layout.width`) and `height` (which should default to `layout.height`).
- `Plotly.toImage` should be able to be in promise chain as such:

``` js
var img = document.createElement('img');
document.body.append(img);

Plotly.newPlot('myDiv', data, layout)
  .then(Plotly.toImage)
  .then(function(url){ img.src = url });
```
- `Plotly.toImage` should be defined in `src/plot_api/to_image.js`
- The `Plotly.Snapshot` should keep the same default behavior as their current version. Unfortunately, these methods are _unofficially_ exposed at the moment, and this PR shouldn't make any breaking changes to the library.

timelyportfolio commented:
I believe I made all the changes mentioned in the line notes.  However, I need to return the `Snapshot` back to the way it was, and move the new `toImage` and the new `downloadImage` to attach directly to `Plotly`.

@etpinard, do you envision a `1.9.1` for this?

timelyportfolio commented:
@etpinard, I'll try to summarize the current status of this pull, since a lot has been circular.
- [ ] still need to add tests, but want to make sure that I understood the requirements first

### `Plotly.Snapshot`

Back to the way it was except
1. `svgToImg` will return an `EventEmitter` or optionally a `Promise` if `opts` contains `promise: true` .
2. We have a new `downloadImage` function that programatically does what the snapshot modebar button does.  `downloadImage` supports multiple formats (`png`, `jpeg`, and `webp`) and also supports `height` and `width`.
3.  Since the snapshot button and `downloadImage` do the same thing, the click handler for the snapshot modebar now just uses `downloadImage`.

### Plotly.toImage

The new promise-based `toImage` is attached directly to `Plotly` and not hidden behind the unofficial `Plotly.Snapshot`.  Note, `Plotly.toImage` is different from `Plotly.Snapshot.toImage`.  We retained the `Plotly.Snapshot.toImage` for backward compatibility in case some users have taken advantage of the functionality in the past.  Plans are to remove the `Plotly.Snaphot` unofficial API in `2.0.0`.

Here is a simple example use of the new `toImage`.

```
var img = document.createElement('img');
document.body.appendChild(img);

Plotly.plot()
    .then(Plotly.toImage)
    .then(function(url){
         img.src = url;
    })
```

... but we also get to specify `format`, `height`, and `width`, which would look like this.

```
var img = document.createElement('img');
document.body.appendChild(img);

Plotly.plot()
    .then(function(gd){
         return Plotly.toImage(gd,{format:'webp', height:400, width:400});
    })
    .then(function(url){
         img.src = url;
    })
```

etpinard commented:
@timelyportfolio you're on the right track for sure. :+1: for points 1 to 3.

As `Plotly.toImage` will be part of our official API, we should make sure that it handles bad arguments in a sensible way. Looking over the [`plot_api_test.js`](https://github.com/plotly/plotly.js/blob/master/test/jasmine/tests/plot_api_test.js#L71) suite is a good place to start.

I'll make a few styling comments on the modified lines directly.

etpinard commented on src/components/modebar/buttons.js:
```diff
@@ -47,50 +47,8 @@ modeBarButtons.toImage = {
-
-            ev.clean();
-        });
+    click: function(gd){
```
`function(gd) {`

etpinard commented on src/plot_api/to_image.js:
```diff
@@ -0,0 +1,95 @@
+ * @param opts.format 'jpeg' | 'png' | 'webp' | 'svg'
+ */
+function toImage(gd, opts) {
+    var Snapshot = require('../snapshot');
```
bring `var Snapshot = require('../snapshot');` just below `var Plotly = require('../plotly');`

timelyportfolio replied:
embarrassed to say, but when I do this `Snapshot` is empty, and I don't understand why.  What I understand about scoping says that this makes no sense.  Defining as before and where you suggest should make it available within, but for some reason it is an empty `{}` when I put it where it belongs.

etpinard replied:
Oh. That mean there's a circular dependency somewhere between `src/plot_api/` and `src/snapshot`. 

Removing that circular dependency might be tricky. I think it would require splitting up `src/plot_api/plot_api.js` into smaller files. Something nobody really wants to do.

You should be able to get away with using `Plotly.Snapshot` in `src/plot_api/to_image.js` without any issues.

timelyportfolio replied:
ok

etpinard commented on src/plot_api/to_image.js:
```diff
@@ -0,0 +1,95 @@
+/**
+ * @param {object} gd figure Object
+ * @param {object} opts option object
+ * @param opts.format 'jpeg' | 'png' | 'webp' | 'svg'
```
add info about `opts.width` and `opts.height`

etpinard commented on src/plot_api/to_image.js:
```diff
@@ -0,0 +1,95 @@
+                        //  rather than EventEmitter
+                        //  leave EventEmitter for backward
+                        //  compatibility
+                        promise: true
```
nicely done.

etpinard commented on src/plotly.js:
```diff
@@ -79,6 +79,12 @@ exports.register(require('./traces/scatter'));
+// toImage to attach directly to Plotly
+//  to allow us to get rid of Snapshot below
+//  for version 2.0.0
+exports.toImage = require('./plot_api/to_image');
```
no need for this line.

Your patch in `src/core.js` is sufficient to expose `Plotly.toImage`.

timelyportfolio replied:
If read correctly, then if I eliminate this line, `Plotly.toImage` will be undefined, since I do not attach `toImage` in [`plot_api.js`](https://github.com/timelyportfolio/plotly.js/blob/master/src/plot_api/plot_api.js).  I'm happy to add this in `plot_api.js`, but I wanted to confirm before doing.

timelyportfolio replied:
Never mind, I can just do the `require` in `core.js`.

etpinard commented on src/snapshot/download.js:
```diff
@@ -0,0 +1,66 @@
+    opts.format = opts.format || 'png';
+    
+    if(Lib.isIE()) {
+        Lib.notifier('Snapshotting is unavailable in Internet Explorer. ' +
```
We make this an option.

I think the mode bar `toImage` button should use `Lib.notifier` to display messages (as currently) whereas `Plotly.downloadImage` method should throw `new Error`s.

etpinard commented on src/snapshot/download.js:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    if(gd._snapshotInProgress) {
+        Lib.notifier('Snapshotting is still in progress - please hold', 'long');
```
Similarly, `Plotly.downloadImage` should throw an Error here.

etpinard commented on src/snapshot/download.js:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    gd._snapshotInProgress = true;
+    Lib.notifier('Taking snapshot - this may take a few seconds', 'long');
```
this notifier should be omitted in `Plotly.downloadImage`.

etpinard commented on src/snapshot/download.js:
```diff
@@ -0,0 +1,66 @@
+
+    var promise = toImage(gd, opts);
+
+    var filename = gd.fn || 'newplot';
```
we should add an `opts` field for the filename e.g.

``` js
Plotly.plot(gd, data, layout).then(function(gd) {
  return Plotly.downloadImage(gd, { filename: 'my-graph', format: 'png' });
});
```

etpinard commented on src/snapshot/svgtoimg.js:
```diff
@@ -6,64 +6,87 @@
+            // eventually remove the ev
+            //  in favor of promises
+            if(!opts.promise){
+              return ev.emit('error', err);
```
nicely done.

etpinard commented on src/snapshot/download.js:
```diff
@@ -25,41 +27,30 @@ function downloadImage(gd, opts) {
     var promise = toImage(gd, opts);
 
-    var filename = gd.fn || 'newplot';
+    var filename = opts.filename || gd.fn || 'newplot';
```
nice :+1:

etpinard commented on src/core.js:
```diff
@@ -31,7 +31,7 @@ exports.moveTraces = Plotly.moveTraces;
 exports.setPlotConfig = require('./plot_api/set_plot_config');
 exports.register = Plotly.register;
-exports.toImage = Plotly.toImage;
+exports.toImage = require('./plot_api/to_image');
```
nice :+1:

etpinard commented on src/plot_api/to_image.js:
```diff
@@ -27,6 +27,12 @@ function toImage(gd, opts) {
         // default to png
         opts.format = opts.format || 'png';
+        
+        if( (opts.width && opts.width < 1) ||
```
let's be even more strict by using `var isNumeric = require('fast-isnumeric');` 

``` js
(!isNumeric(opts.width) && opts.width < 1) // ...
```

etpinard replied:
and similarly for `opts.height`

etpinard commented on src/snapshot/download.js:
```diff
@@ -27,13 +27,18 @@ function downloadImage(gd, opts) {
-
     return new Promise(function(resolve,reject){
+        if(gd._snapshotInProgress){
+          reject('Snapshotting is unavailable in Internet Explorer. ' +
```
we should reject a `new Error` instead of just a string. See [example](https://github.com/plotly/plotly.js/blob/master/src/plots/geo/geo.js#L113-L118).

etpinard commented on test/jasmine/tests/toimage_test.js:
```diff
@@ -34,4 +33,79 @@ describe('Plotly.toImage', function() {
+                    expect(img.height).toBe(600);
+                    expect(img.width).toBe(700);
+                    // now provide height and width in opts
+                    Plotly.toImage(gd, {height: 400, width: 400}).then(function(url){
```
could you rewrite this as a promise chain:

``` js
Plotly.plot(/* */).then(function() {

  // ...

 return Plotly.toImage(/* */);
}).then(function() {

  // ...

  return Plotly.toImage(/* */);
}).then(function() {

  // ...

  done();
};
```

to keep things flat.

etpinard commented on test/jasmine/tests/toimage_test.js:
```diff
@@ -34,4 +33,79 @@ describe('Plotly.toImage', function() {
+                        expect(url.substr(0,16)).toBe('data:image/webp;');
+                        Plotly.toImage(gd, {format: 'svg'}).then(function(url){
+                            expect(url.substr(1,3)).toBe('svg');
+                            done();
```
and similarly here.

etpinard commented on test/jasmine/tests/toimage_test.js:
```diff
@@ -34,4 +33,79 @@ describe('Plotly.toImage', function() {
+            
+    });
+    
+    it('should create proper file type', function(done){
```
great tests :tada:

etpinard commented on src/plotly.js:
```diff
@@ -82,3 +82,4 @@ exports.PlotSchema = require('./plot_api/plot_schema');
 
 // imaging routines
 exports.Snapshot = require('./snapshot');
+
```
:cow2: no need for this blank line.

etpinard commented on src/core.js:
```diff
@@ -31,6 +31,7 @@ exports.moveTraces = Plotly.moveTraces;
 exports.purge = Plotly.purge;
 exports.setPlotConfig = require('./plot_api/set_plot_config');
 exports.register = Plotly.register;
+exports.toImage = require('./plot_api/to_image');
```
@timelyportfolio any objections to add `downloadImage` to `Plotly` in this PR?

etpinard replied:
the way you wrote `snapshot/download.js`, I think it would as easy as adding:

``` js
exports.downloadImage = require('./snapshot/download');
```

below this line.

timelyportfolio replied:
makes sense, and I am sure it would get used.  I'll just need to add some tests to feel good about it if it is an "official" function.

timelyportfolio commented:
I rebased to the current master and squashed all the commits into one in this [`feature/snapshot-pull`](https://github.com/timelyportfolio/plotly.js/commits/feature/snapshot-pull) branch.  I'll try to think of how best to test `downloadImage`, rebase/squash again, and then submit a proper pull.

timelyportfolio commented:
@etpinard, I just made some tests for `downloadImage`, but unfortunately, the really negative side effect is we get a download image on the local machine each time the test runs, so if watching, then you might get >10 new files downloaded to your drive.  Any thoughts on whether or not to include?  I condensed the tests so that we only run `downloadImage` once.

```
var Plotly = require('@lib/index');
var createGraphDiv = require('../assets/create_graph_div');
var destroyGraphDiv = require('../assets/destroy_graph_div');
var textchartMock = require('../../image/mocks/text_chart_arrays.json');

// minimize tests since each downloadImage
//  will result in a new file on the machine
//  this might mean we actually just do not test

describe('Plotly.downloadImage', function() {
    'use strict';
    var gd;
    var originalTimeout;

    beforeEach(function() {
        gd = createGraphDiv();

        // downloadImage can take a little longer
        //  so give it a little more time to finish
        originalTimeout = jasmine.DEFAULT_TIMEOUT_INTERVAL;
        jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000;
    });

    afterEach(function() {
        destroyGraphDiv();
        jasmine.DEFAULT_TIMEOUT_INTERVAL = originalTimeout;    
    });

    it('should be attached to Plotly', function() {
        expect(Plotly.downloadImage).toBeDefined();
    });

    it('should create link, remove link, accept options', function(done) {
        //use MutationObserver to monitor the DOM
        //for changes
        //code modeled after
        //https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver
        // select the target node
        var target = document.body;
        var domchanges = [];

        // create an observer instance
        var observer = new MutationObserver(function(mutations) {
          mutations.forEach(function(mutation) {
            domchanges.push(mutation);
          });
        });

        Plotly.plot(gd, textchartMock.data, textchartMock.layout).then(function(gd) {
            // start observing dom
            // configuration of the observer:
            var config = { childList: true };

            // pass in the target node and observer options
            observer.observe(target, config);

            return Plotly.downloadImage(gd, {format: 'jpeg', height: 300, width: 300, filename: 'plotly_download'});
        }).then(function(){
            // stop observing
            observer.disconnect();
            // look for an added and removed link
            var linkadded = domchanges[domchanges.length-2].addedNodes[0].outerHTML;
            var linkdeleted = domchanges[domchanges.length-1].removedNodes[0].outerHTML;

            // check for a <a element and proper file type
            expect(linkadded.substr(0,25)).toBe('<a href="data:image/jpeg;');
            // check that filename option handled properly
            expect(linkadded.substr(linkadded.length-26,15)).toBe('plotly_download');

            // check that link removed
            expect(linkadded).toBe(linkdeleted);
            done();
        });
    });
});

```

etpinard commented:
@timelyportfolio to ensure that we test everything about `downloadImage` except the click call, we could add in the `downloadImage` suite:

``` js

var createElement = document.createElement;

beforeAll(function() {
  document.createElement = function(args) {
     var el = createElement.call(document, args);
     el.click = function() {};
     return el;
  }
});

afterAll(function() {
  document.createElement = createElement;
});
```

timelyportfolio commented:
close in favor of https://github.com/plotly/plotly.js/pull/446
